packages/react-dom/src/__tests__/ReactDOMInput-test.js

```diff
@@ -139,72 +147,95 @@ describe('ReactDOMInput', () => {
     );
   });
 
-  it('should warn for controlled value of false with missing onChange', () => {
-    expect(() =>
-      ReactDOM.render(<input type="checkbox" checked={false} />, container),
-    ).toErrorDev(
+  it('should warn for controlled value of false with missing onChange', async () => {
+    await expect(async () => {
+      await act(() => {
+        root.render(<input type="checkbox" checked={false} />);
+      });
+    }).toErrorDev(
       'Warning: You provided a `checked` prop to a form field without an `onChange` handler.',
     );
   });
 
-  it('should warn with checked and no onChange handler with readOnly specified', () => {
-    ReactDOM.render(
-      <input type="checkbox" checked={false} readOnly={true} />,
-      container,
-    );
-    ReactDOM.unmountComponentAtNode(container);
+  it('should warn with checked and no onChange handler with readOnly specified', async () => {
+    await act(() => {
+      root.render(<input type="checkbox" checked={false} readOnly={true} />);
+    });
+    root.unmount();
+    root = ReactDOMClient.createRoot(container);
 
-    expect(() =>
-      ReactDOM.render(
-        <input type="checkbox" checked={false} readOnly={false} />,
-        container,
-      ),
-    ).toErrorDev(
+    await expect(async () => {
+      await act(() => {
+        root.render(<input type="checkbox" checked={false} readOnly={false} />);
+      });
+    }).toErrorDev(
       'Warning: You provided a `checked` prop to a form field without an `onChange` handler. ' +
         'This will render a read-only field. If the field should be mutable use `defaultChecked`. ' +
         'Otherwise, set either `onChange` or `readOnly`.',
     );
   });
 
-  it('should not warn about missing onChange in uncontrolled inputs', () => {
-    ReactDOM.render(<input />, container);
-    ReactDOM.unmountComponentAtNode(container);
-    ReactDOM.render(<input value={undefined} />, container);
-    ReactDOM.unmountComponentAtNode(container);
-    ReactDOM.render(<input type="text" />, container);
-    ReactDOM.unmountComponentAtNode(container);
-    ReactDOM.render(<input type="text" value={undefined} />, container);
-    ReactDOM.unmountComponentAtNode(container);
-    ReactDOM.render(<input type="checkbox" />, container);
-    ReactDOM.unmountComponentAtNode(container);
-    ReactDOM.render(<input type="checkbox" checked={undefined} />, container);
+  it('should not warn about missing onChange in uncontrolled inputs', async () => {
```

Comment:
Ditto here, could probably use a key.

Reply:
Since this contains so many changes already, I'm trying to keep the changes minimal. Refactoring some tests could still be done as a PR on top of this.